src/components/ec/cuda/kernel/ec_cuda_executor.cu

```diff
@@ -0,0 +1,152 @@
+{
+    *state = UCC_EC_CUDA_EXECUTOR_SHUTDOWN_ACK;
+}
+
+template <typename T>
+__device__ void executor_copy(T* __restrict__ d, T* __restrict__ s,
+                              size_t count)
+{
+    size_t start = threadIdx.x;
+    const size_t step  = blockDim.x;
+
+    for (size_t i = start; i < count; i+=step) {
+        d[i] = s[i];
+    }
+}
+
+template <typename T>
+__device__ void executor_copy_aligned(T* __restrict__ d, T* __restrict__ s,
+                                      size_t count)
+{
+    size_t idx = threadIdx.x;
+    const size_t step  = blockDim.x;
+    const int n = count / sizeof(T);
+    const int num_iter = n / step + ((idx < n % step) ? 1 : 0);
+    char1 *s1 = (char1*)s;
+    char1 *d1 = (char1*)d;
+
+#pragma unroll 4
+    for(int i = 0; i < num_iter; i++) {
+        d[i * step + idx] = s[i * step + idx];
+    }
+
+    if (idx < count % sizeof(T)) {
+        d1[count - idx - 1] = s1[count - idx - 1];
+    }
+}
+
+__global__ void executor_kernel(volatile ucc_ec_cuda_executor_t *eee,
+                                int q_size)
+{
+    const uint32_t  worker_id   = blockIdx.x;
+    const uint32_t  num_workers = gridDim.x;
+    bool            is_master   = (threadIdx.x == 0) ? true: false;
+    int             cidx_local, pidx_local;
+    volatile int *pidx, *cidx;
+    ucc_ee_executor_task_t *tasks;
+    __shared__ ucc_ee_executor_task_args_t args;
+    __shared__ bool worker_done;
+
+    if (is_master) {
+        cidx_local = worker_id;
+        pidx       = eee->dev_pidx;
+        cidx       = eee->dev_cidx;
+        tasks      = eee->dev_tasks;
+    }
+
+    worker_done = false;
+    __syncthreads();
+    while (1) {
+        if (is_master) {
+            while ((*cidx % num_workers) != worker_id);
+            do {
+                pidx_local = *pidx;
+            } while (*cidx == pidx_local);
+            (*cidx)++;
+            worker_done = (pidx_local == -1);
+            if (!worker_done) {
+                args = tasks[cidx_local].args;
+            }
+        }
+        __syncthreads();
+        if (worker_done) {
+            return;
+        }
+        switch (args.task_type) {
+            bool aligned;
+            case UCC_EE_EXECUTOR_TASK_TYPE_COPY:
+                aligned = !(align_pow2((intptr_t)args.bufs[0], 16) ||
+                            align_pow2((intptr_t)args.bufs[1], 16));
+                if (aligned) {
+                    executor_copy_aligned<uint4>((uint4*)args.bufs[0],
+                                                 (uint4*)args.bufs[1],
+                                                 args.count);
+
+                } else {
+                    executor_copy((char*)args.bufs[0],
+                                  (char*)args.bufs[1],
+                                   args.count);
+                }
+                break;
+            default: break;
+        }
+        __syncthreads();
+        __threadfence_system();
```

Comment:
why need threadfence_system ?

Reply:
__threadfence_system() is needed to make sure nvlink transfer is completed